### ISSUE TITLE
Metadata coherence AccesMode

### ISSUE THREAD

**Issue by laudrain** [daisy/ace#199]
Please provide the following details if possible or relevant:

- [1.0.2] the version of Ace impacted by the issue (as returned by the `ace -v` command).
- [v8.7.0] the version of Node used to run Ace (as returned by the `node -v` command).
- [MacOS 10.3.6] your operating system and version.
- [] if the issue is a bug when running Ace: the error message as produced in the log file (you can enable debug logging with the `--verbose` option).
- [] if the issue is about the HTML report: the browser(s) and version(s) you used.
- [] if possible, let us know if you can share a sample input file to help us reproduce the issue.

Ace should verify coherence between accesMode and EPUB content.
If an EPUB has images, an accessMode visual must be present in the metadata.
`<meta property="schema:accessMode">visual</meta>`

**Comment by clapierre** [comment]
I was thinking the same thing, there shouldn’t be accessMode Visual in a purely decorative image book.
Therefore wouldn’t accessMode Visual be a violation if accessMode Visual is stated in a book without images or are all marked decorative?

The only exception to this would be an image for the cover page I would argue.

Thanks
EOM

Charles LaPierre
Technical Lead, DIAGRAM and Born Accessible
E-mail: charlesl@benetech.org<mailto:charlesl@benetech.org>
Twitter: @CLaPierreA11Y
Skype: charles_lapierre
Phone: 650-600-3301



On Jul 20, 2018, at 5:21 AM, Romain Deltour <notifications@github.com<mailto:notifications@github.com>> wrote:


Ace should verify coherence between accesMode and EPUB content.
If an EPUB has images, an accessMode visual must be present in the metadata.

Unfortunately it’s not that simple: if the images are purely decorative, then "visual" isn’t an access mode.

I agree that some things could be detected with better heuristics however, for instance when images with alt text are detected.

**Comment by laudrain** [comment]
If I understand correctly, accessMode metadata is supposed to list all "human sense perceptual system or cognitive faculty through which a user may process or perceive the content of a digital resource." 
> if an EPUB Publication contains images and video, visual perception is required to consume the content exactly as it was created.

in http://www.idpf.org/epub/a11y/techniques/#meta-001
At this stage, I believe it's not a question of adaptation, which is then treated in accessModeSufficent.

**Comment by clapierre** [comment]
I think an argument can be made either way  -- if all images are decorative it is definitely true that there should be a line for accessModeSufficient=textual. But it isn't entirely untrue to also say accessMode=visual since there are images to look at. They just aren't mandatory for understanding the book. The same is true if all images have alt tags – there are images, but they aren't needed to understand the book because there is text to replace them.

I think the difference might be (if I understand the task you are carrying out, which I may not) that you can assess through the metadata if the book has alt tags, but you cannot assess if images are present but tagged decorative?

Since you are trying to create a set of reliable algorithms I would err on the side of simplicity, so choose whichever will give the tool more consistent and useful output.

(Charles, if this doesn't post to the lists because I am not a member, please forward for me.)

-Madeleine

From: Charles LaPierre <charlesl@benetech.org>
Date: Friday, July 20, 2018 at 8:42 AM
To: daisy/ace <reply@reply.github.com>
Cc: daisy/ace <ace@noreply.github.com>, Subscribed <subscribed@noreply.github.com>, Madeleine Rothberg <madeleine_rothberg@wgbh.org>
Subject: Re: [daisy/ace] Metadata coherence AccesMode (#199)

I was thinking the same thing, there shouldn’t be accessMode Visual in a purely decorative image book.
Therefore wouldn’t accessMode Visual be a violation if accessMode Visual is stated in a book without images or are all marked decorative?

The only exception to this would be an image for the cover page I would argue.

Thanks
EOM

Charles LaPierre
Technical Lead, DIAGRAM and Born Accessible
E-mail: charlesl@benetech.org<mailto:charlesl@benetech.org>
Twitter: @CLaPierreA11Y
Skype: charles_lapierre
Phone: 650-600-3301



On Jul 20, 2018, at 5:21 AM, Romain Deltour <notifications@github.com<mailto:notifications@github.com>> wrote:

Ace should verify coherence between accesMode and EPUB content.
If an EPUB has images, an accessMode visual must be present in the metadata.
Unfortunately it’s not that simple: if the images are purely decorative, then "visual" isn’t an access mode.
I agree that some things could be detected with better heuristics however, for instance when images with alt text are detected.

**Comment by laudrain** [comment]
Sorry but I am not speaking of accessModeSufficient, but on accessMode only.
In schema.org, [accessMode](https://schema.org/accessMode) is neutral on adaptation or presentation, it's only perception: 

> The human sensory perceptual system or cognitive faculty through which a person may process or perceive information.

**Comment by clapierre** [comment]
After hearing from Luc and Madeleine, I now agree that 
accessMode Visual is valid for images marked presentational as it is a mode that you can perceive the information.  

Actually accessModeSufficient visual/textual is really splitting hairs because you can consume the information  either all visually or textually  so having the combination is probably a judgement call, I can see both arguments we would need a rule that says something like, this can only be applied together if both separate access modes are not present meaning that there may be some images that are required that are not described.

**Comment by rdeltour** [comment]
I believe it depends on the use case:
 * on the one hand, the use of images as bullet list items (for whatever design decision), or the insertion of just the publishers’s logo, don't justify declaring "visual"  as an access mode.
 * On the other hand, "decorative" images can play a real and important part in the visual design, while not adding any useful content, in which case I would agree that a "visual" access mode is legitimate.

In other words, as often with accessibility, checking the accessmode coherence requires human inspection.

That said, it's certainly possible to check _some_ things, in _some_ cases. We need to figure out the right algorithm, and describing it as an accessibility test rule.

**Comment by laudrain** [comment]
Sorry, but I still disagree: i think accessMode has nothing to do with usefulness of image content.
I would say that the presence of at least one image, be it the publisher's logo, needs to be declared with accessMode visual.

**Comment by laudrain** [comment]
I would say the same for auditory accessMode.
Even if the sound is a purely decorative jingle, the presence of one sound justifies the need of auditory accessMode.

**Comment by clapierre** [comment]
I wasn't thinking of it that way Luc, but you do have a point, and really accessModeSufficient is where we need to be looking to see if I can sufficiently access the material in the modalities which I can perceive.

So in that case it makes accessMode really easy to check.  If there are any images in the book "visual" gets added, "textual" would be pretty much in every book unless its an audio book for example, "audio" would appear if there are any audio in the book at all.
Where it becomes tricky is the accessModeSufficient.

**Comment by mattgarrish** [comment]
This is one of the problems of accessMode and why we also have accessModeSufficient. Listing the accessModes doesn't indicate whether the content is significant or not. That's where accessModeSufficient comes in to tell you whether you actually need to be able to perceive the visual content. To quote from the wiki:

> The main drawback of using accessMode in our context is that it does not describe to the primary mode of access. For instance, a text-based publication which happen to also contain a few images and one audio sample would have an accessMode property with values textual, visual, and auditory, with no indication of which mode is primary, or whether the entire content is accessible with a certain mode.

**Comment by clapierre** [comment]
accessModePrimary ummm thats interesting addition.  I wonder if that would be something the EPUB CG would want to take up.
Since accessModeSufficient doesn't really cover what would be the primary modality you would need to use to access the majority of the book.  Would be helpful to know if this is more of a visual book, audio book or textual book with other bits thrown in.

**Comment by laudrain** [comment]
@clapierre Yes accessModeSufficient is tricky, but that's another question. I have brought other issues on that..

**Sticking to accessMode**, we should not take in consideration if "the content is significant or not"(thanks to @mattgarrish).
And I completely agree with the wiki excerpt @mattgarrish has given:
> For instance, a text-based publication which happen to also contain a few images and one audio sample would have an accessMode property with values textual, visual, and auditory

No need to say primary here, the schema.org is clear enough:
> The human sensory perceptual system or cognitive faculty through which a person may process or perceive information.

**Comment by rdeltour** [comment]
@laudrain & @mattgarrish 
> Listing the accessModes doesn't indicate whether the content is significant or not.

I disagree. Of course, I see where you come from, and I would agree most of the time. But since we’re in the Ace issue tracker, we’re talking about the possibility for _automated_ checking, and we have to consider edge cases; I simply don't think that checking whether a "visual" access mode is required boils down to answering the question "does the content have at least one visible `img` element".

What if the image is a spacer gif? What if it is a tracking pixel? What if the image is just used to add a border?
These things exist in the real world, and it’s hard to tell with automated checking.

If you argue the things above warrant declaring a visual access mode, then the exact same visual effect achieved through CSS does as well. Does it mean we need to declare a "visual" access mode whenever a border property is set in CSS? I don't think so. I believe that declaring whether the content is accessed visually or not sometimes require some human judgment; the line between visual or not can be thin, and I don't think the automated tool can make a definitive call.

**Comment by laudrain** [comment]
In EPUB world, I don't believe tracking pixels and spacer gif do exist really.
Except if Reading Systems add this kind of content in EPUBs after the publisher delivery.
But even then, it is not in the scope of an accessibility checker for EPUB.

And a border is more on the textual side, as any text decoration.

To stay "on the side of simplicity" (from Madeleine above comment), IMO automated control can even be made on the OPF itself as all resources are declared in the manifest.

**Comment by rdeltour** [comment]
> In EPUB world, I don't believe tracking pixels and spacer gif do exist really.
> Except if Reading Systems add this kind of content in EPUBs after the publisher delivery.
> But even then, it is not in the scope of an accessibility checker for EPUB.
> 

I agree that tracking pixels shouldn't appear much in EPUB, as it wouldn't pass EpubCheck (remote resource).

I've seen things like borders and other "design-only" images however, although I don't have stats of real-world usage.

> And a border is more on the textual side, as any text decoration.

So you agree an image used for a border doesn't warrant the "visual" access mode, do I understand correctly?

> To stay "on the side of simplicity" (from Madeleine above comment), IMO automated control can even be made on the OPF itself as all resources are declared in the manifest.

Images could be declared in the OPF but not actually used in content…

**Comment by mattgarrish** [comment]
> I've seen things like borders and other "design-only" images however, although I don't have stats of real-world usage.

This is really getting into the weeds, but  I agree with you that in reality these purely decorative design bits are fluff and don't warrant a visual access mode. If someone lists them, though, they're also not going to be harmful provided the sufficientAccessModes property disregards them.

It does get silly if we say a book has a visual access mode because there are decorative curlicues around the chapter numbers.

When I think of decorative images being counted in the access modes, my mind is more on images that might adorn a new section, but are generally irrelevant (e.g., full-page cityscapes, nature scenes, etc. that maybe give a bit of mood to a section but aren't actually conveying any meaningful information).

As far as ace checking goes, I think it might be worth questioning whether a visual access mode is necessary if there are images, but I'm not suggesting that Ace will ever get this right on its own or that it should throw an error if the metadata is not present. We should always defer to human interpretation on both accessmode and accessmodesufficient.

**Comment by laudrain** [comment]
> So you agree an image used for a border doesn't warrant the "visual" access mode, do I understand correctly?

To keep it simple, If an image is used for a border, it is an image, then it needs visual. 
In that case the CSS will call for an image file (any format, even SVG) and that file will be declared in the manifest.

> Images could be declared in the OPF but not actually used in content…

That's a useless and bad practices. I would hope epubcheck to flag this.
At least, Ace asking for a accessMode visual will reduce this errand though I doubt there are EPUB without any image as the cover is a must have.

**Comment by laudrain** [comment]
> We should always defer to human interpretation on both accessmode and accessmodesufficient.

@mattgarrish I completely agree on accessModeSuficient, but not on accessMode.
We need a strong base on accessMode to build intelligible messages to the community, particularly for EPUB producers.
And it will then be _a bit_ more easy to explain accessModeSufficent !

**Comment by clapierre** [comment]
I think this is a judgement call as there are too many fringe cases and lets leave it to the Certifier to determine how they interpret what is visual and what is just purely decorative and should not justify that accesscode being present.  As long as that Certifier is consistent.  

For our Global Certified Accessible program I will be referencing this issue when we need to make a judgement call whether or not to include the various accessModes when there is some question.
Great discussion thanks!

**Comment by rdeltour** [comment]
> To keep it simple, If an image is used for a border, it is an image, then it needs visual.

Let's agree to disagree 🙂

>> Images could be declared in the OPF but not actually used in content…
> That's a useless and bad practices. I would hope epubcheck to flag this.

Unfortunately here again there's not much EpubCheck can do, as we can't easily and predictably know if a resource will be used by dynamically-generated content (JS).

**Comment by rdeltour** [comment]
> If Ace doesn’t do that check, we will do it on our side.

With a future version of Ace, you would implement it as a custom check :-)

**Comment by rdeltour** [comment]
> I conclude for myself that an automated check can be done between image presence in the OPF manifest declarations and the accessMode visual in metadata.
> If Ace doesn’t do that check, we will do it on our side.
> This will help the real world of ePub producers to cope more easily with implementing accessibility features in their process.

That's the thing: if you have some control on the production process, or want to provide guidelines, then it makes perfect sense to require and expect that all images translate in a "visual" access mode. Do it by all means!

But for Ace generally, we have to support all the content that is produced in the wild world. If we want to avoid false positives, as an automated testing tool should do IMO, then we don't have a choice but let that pass and defer to a human's judgment.

BTW, [I just asked #eprdctn on Twitter](https://twitter.com/rdeltour/status/1020318561439559682) and got confirmation that while it's knowingly considered a bad practice, hacks relying on non-visually-relevant images _are_ used.

Anyways, interesting discussion 👍
